system/jlib/jsocket.cpp

```diff
@@ -5103,63 +5129,128 @@ class CSocketEpollThread: public CSocketBaseThread
             EXCLOG(e,"CSocketEpollThread");
             termexcept.setown(e);
         }
+        finished = true;
         return 0;
     }
 };
 
 class CSocketEpollHandler: implements ISocketSelectHandler, public CInterface
 {
-    CSocketEpollThread *epollthread;
+    std::list<Owned<CSocketEpollThread>> threads;
     CriticalSection sect;
+    bool started;
     StringAttr epolltrace;
+    unsigned hdlPerThrd;
 public:
     IMPLEMENT_IINTERFACE;
-    CSocketEpollHandler(const char *trc)
-        : epolltrace(trc)
+    CSocketEpollHandler(const char *trc, unsigned _hdlPerThrd)
+        : started(false), epolltrace(trc), hdlPerThrd(_hdlPerThrd)
     {
-        epollthread = new CSocketEpollThread(epolltrace);
     }
 
     ~CSocketEpollHandler()
     {
-        delete epollthread;
+        stop(true);
+        threads.clear();
     }
 
     void start()
     {
         CriticalBlock block(sect);
-        epollthread->start();
+        if (!started)
+        {
+            started = true;
+            std::list<Owned<CSocketEpollThread>>::iterator it = threads.begin();
+            while (it != threads.end())
+            {
+                (*it)->start();
+                it++;
+            }
+        }
     }
 
     void add(ISocket *sock,unsigned mode,ISocketSelectNotify *nfy)
     {
-        CriticalBlock block(sect); // JCS->MK - are these blocks necessary? epollthread->add() uses it's own CS.
+        if ( !sock || !nfy ||
+             !(mode & (SELECTMODE_READ|SELECTMODE_WRITE|SELECTMODE_EXCEPT)) )
+            throw MakeStringException(-1,"CSocketEpollHandler::add() invalid sock or nfy or mode");
+
+        CriticalBlock block(sect);
+        // Create new handler thread if current one has hdlPerThrd fds.
+        // epoll() handles many fds faster than select so this would
+        // seem not as important, but we are still serializing on
+        // nfy events and spreading those over threads may help,
+        // especially with SSL as avail_read() could block more.
+        unsigned nt = threads.size();
+        bool added = false;
+        bool removed = false;
+        std::list<Owned<CSocketEpollThread>>::iterator it = threads.begin();
+        while (it != threads.end())
+        {
+            if (added)
+            {
+                if (!removed)
+                {
+                    if ((*it)->remove(sock, nt) && sock)
```

Comment:
sock cannot be null since tested at the head of the function

Reply:
ok, yes, removed extra check